### PR TITLE
[Article] Vue3 composition api in a nutshell

### PR DESCRIPTION
**Pull request by larsjanssen6**

### CODE REVIEW

File: data/blog/vue3-composition-api-in-a-nutshell.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+title: 'Vue 3 composition API in a nutshell'
+date: '2022-10-25'
+tags: ['developtment', 'frontend', 'vue']
+images: ['/articles/vue3-composition-api-in-a-nutshell/header.png']
+summary: ''
+authors: ['lars-janssen']
+theme: 'black'
+---
+
+Vue 3 has been released on Monday, February 7, 2022. Since the release of vue 3 the ecosystem got a boost (with for example Vite) and
```
Comment (mrtnvh) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: Typo. Vue with a capital V.

File: data/blog/vue3-composition-api-in-a-nutshell.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+the options API has been introduced. In this article I will explain some core concepts of the options API.
+
+## What's the composition API?
+
+With the arrival of Vue 3 we can declare components with the `options API` (old way) or the `composition API` (new way). The composition API is a set of APIs that allows us to author Vue components using imported functions instead of declaring options.
+
+With the composition API the vue components are more structured and readable, this can be very helful especially when components grow. The following image displays more structured components.
```
Comment (mrtnvh) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: Typo. Vue with a capital V.

File: data/blog/vue3-composition-api-in-a-nutshell.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+![vue3-options-api-composition-api](/articles/vue3-composition-api-in-a-nutshell/options-api-composition-api.png)
+
+According to the official vue documentation the most important benefits are:
+
+- Better logic Reuse;
```
Comment (mrtnvh) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: Should Reuse be capitalized?

File: data/blog/vue3-composition-api-in-a-nutshell.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+- Better type inference;
+- Smaller production bundle and less overhead.
+
+### Setup
+
+With the composition API we can declare our vue template in 2 different ways:
+
+Method 1:
+
+```js
```
Comment (mrtnvh) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: Make 3 lines of this in favour of one line?

File: data/blog/vue3-composition-api-in-a-nutshell.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+const dataProperty: string = ref('')
+
+watch(() => {
+  //Watch data
+})
+```
+
+### Props
+
+Vue 3 has been rewritten in typescript. We notice this well as typescript can be better integrated into our application. Props are a good example of this. With the options API:
```
Comment (mrtnvh) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: TypeScript

File: data/blog/vue3-composition-api-in-a-nutshell.md [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+```js
+props: {
+    dogNames: {
+        type: Array,
+        default: []
+    }
+}
+```
+
+In vue 2 you could already define a type, but this was still quite limited. You now have the option to use interfaces. With the composition API:
```
Comment (mrtnvh) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: Vue

File: data/blog/vue3-composition-api-in-a-nutshell.md [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+let props = withDefaults(defineProps<Props>(), {
+  dogNames: [],
+});
+```
+
+### Events
+
+When dispatching events to our parent components we notice the increased typescript support aswel. Types
```
Comment (mrtnvh) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: - TypeScript
- as well

File: data/blog/vue3-composition-api-in-a-nutshell.md [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+With the options API:
+
+```js
+$emit('dogNames', ['Snuffel'])
+```
+
+### Multiple v-models
+
+One of the most requested features was multiple `v-models`, with vue3 it's finally there. Before we had to pass additional
```
Comment (mrtnvh) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: Vue 3

File: data/blog/vue3-composition-api-in-a-nutshell.md [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+```js
+const emit = defineEmits<{
+  (event: "update:dogNames", value: Array<string>): void;
+  (event: "update:catNames", value: Array<string>): void;
+}>();
+```
+
+### Final thoughts
+
+Vue 3 has been a huge update compared to Vue 2. You could almost see it as a new framework rather than an update. Within IO, a number of projects are now done in vue3, in which I participate. In the beginning I had to get used to the new structure of vue and the composition API. Now I see the benefits more and more, and understand why this update was necessary. Especially the typescript integration is very good, and with hooks we reuse our code more easily and is it more readable.
```
Comment (mrtnvh) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: Vue 3

**Review by mrtnvh — CHANGES_REQUESTED** [review]
Hi Lars, welcome to the iO TechHub!

Yes, also here I can review your code. 😄 

Article is looking good! Discovered some typos whilst browsing. Could you change these, perhaps run the Markdown through a Grammarly or other spell checks?

File: data/blog/vue3-composition-api-in-a-nutshell.md [hunk 10]
```diff
@@ -0,0 +1,164 @@
+---
+title: 'Vue 3 composition API in a nutshell'
+date: '2022-10-25'
+tags: ['developtment', 'frontend', 'vue']
```
Comment (mrtnvh) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: development

File: data/blog/vue3-composition-api-in-a-nutshell.md [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+images: ['/articles/vue3-composition-api-in-a-nutshell/header.png']
+summary: ''
```
Comment (mrtnvh) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: Could you add a summary? This is for SEO.

**Review by mrtnvh — CHANGES_REQUESTED** [review]

File: data/blog/vue3-composition-api-in-a-nutshell.md (hunk 1 above)
Comment (RonaldvdH) [reply data/blog/vue3-composition-api-in-a-nutshell.md]: With for example -> for example with?

File: data/blog/vue3-composition-api-in-a-nutshell.md [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+authors: ['lars-janssen']
+theme: 'black'
+---
+
+Vue 3 has been released on Monday, February 7, 2022. Since the release of vue 3 the ecosystem got a boost (with for example Vite) and
+the options API has been introduced. In this article I will explain some core concepts of the options API.
+
+## What's the composition API?
```
Comment (RonaldvdH) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: What's is more informal talk, for writing a blog i would just write "What is".

File: data/blog/vue3-composition-api-in-a-nutshell.md [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+date: '2022-10-25'
+tags: ['developtment', 'frontend', 'vue']
+images: ['/articles/vue3-composition-api-in-a-nutshell/header.png']
+summary: ''
+authors: ['lars-janssen']
+theme: 'black'
+---
+
+Vue 3 has been released on Monday, February 7, 2022. Since the release of vue 3 the ecosystem got a boost (with for example Vite) and
+the options API has been introduced. In this article I will explain some core concepts of the options API.
```
Comment (RonaldvdH) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: Do you actually mean composition API here twice? as i read later that options API is the old way?

File: data/blog/vue3-composition-api-in-a-nutshell.md (hunk 2 above)
Comment (RonaldvdH) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: helful? :)

File: data/blog/vue3-composition-api-in-a-nutshell.md (hunk 4 above)
Comment (RonaldvdH) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: As it is typescript, does markdown understand that? ```ts?

File: data/blog/vue3-composition-api-in-a-nutshell.md (hunk 5 above)
Comment (RonaldvdH) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: "We notice this well" This feels like a very literal dutch translation. Do you want to convey a message something like "The integration of Vue in our TypeScript application is now more smooth since Vue 3 is rewritten in TypeScript.?

File: data/blog/vue3-composition-api-in-a-nutshell.md (hunk 7 above)
Comment (RonaldvdH) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: is it increased support? or did it go from 0 to actual typescript support? if that is the cause i would not use the word "increased" as it insinuates previous support that got better.

File: data/blog/vue3-composition-api-in-a-nutshell.md [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+emit('dogNames', ['Snuffel']);
+```
+
+With the options API:
+
+```js
+$emit('dogNames', ['Snuffel'])
+```
+
+### Multiple v-models
```
Comment (RonaldvdH) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: This post is about the composition API, Multiple v-models seems like an unrelated feature that came with Vue 3.
If it is not unrelated i would like to read the explanation why multiple v-models are now possible because of the composition API

File: data/blog/vue3-composition-api-in-a-nutshell.md (hunk 9 above)
Comment (RonaldvdH) [inline data/blog/vue3-composition-api-in-a-nutshell.md]: IO -> iO

**Review by RonaldvdH — CHANGES_REQUESTED** [review]

File: data/blog/vue3-composition-api-in-a-nutshell.md (hunk 3 above)
Comment (DaveBitter) [reply data/blog/vue3-composition-api-in-a-nutshell.md]: "Better reusability of logic" sounds a bit nicer IMO

File: data/blog/vue3-composition-api-in-a-nutshell.md (hunk 11 above)
Comment (DaveBitter) [reply data/blog/vue3-composition-api-in-a-nutshell.md]: And it is used in the header of the article page

File: data/blog/vue3-composition-api-in-a-nutshell.md (hunk 4 above)
Comment (DaveBitter) [reply data/blog/vue3-composition-api-in-a-nutshell.md]: you can use ts indeed, there is support in highlighting

**Review by DaveBitter — COMMENTED** [review]
Nice article! Good to have as an internal and external reference!

File: data/blog/vue3-composition-api-in-a-nutshell.md (hunk 14 above)
Comment (larsjanssen6) [reply data/blog/vue3-composition-api-in-a-nutshell.md]: Good one. I had this in there because it works a lot better with events (of the composition) API. I've replaced this with using your own hooks vs mixins. Also very relevant.

**Comment by larsjanssen6** [comment]
Thankyou all for the feedback!

**Comment by memic84** [comment]
Hi Lars, it's a good article. Nice to see some Vue content. 

Last year we published also a Vue composition api article, feel free to re-use any if possible. We won't be migrating this one anymore.

https://www.intracto.com/nl-be/blog/composition-api-de-belangrijkste-nieuwigheid-vue-3
